src/Platform/Microsoft.Testing.Platform/TestHostOrcherstrator/TestHostOrchestratorManager.cs

```diff
@@ -48,4 +49,37 @@ public async Task<TestHostOrchestratorConfiguration> BuildAsync(ServiceProvider
 
         return new TestHostOrchestratorConfiguration([.. orchestrators]);
     }
+
+    public void AddTestHostOrchestratorApplicationLifetime(Func<IServiceProvider, ITestHostOrchestratorApplicationLifetime> testHostOrchestratorApplicationLifetimeFactory)
+    {
+        Guard.NotNull(testHostOrchestratorApplicationLifetimeFactory);
+        _testHostOrchestratorApplicationLifetimeFactories.Add(testHostOrchestratorApplicationLifetimeFactory);
+    }
+
+    internal async Task<ITestHostOrchestratorApplicationLifetime[]> BuildTestHostOrchestratorApplicationLifetimesAsync(ServiceProvider serviceProvider)
+    {
+        List<ITestHostOrchestratorApplicationLifetime> lifetimes = [];
+        foreach (Func<IServiceProvider, ITestHostOrchestratorApplicationLifetime> testHostOrchestratorApplicationLifetimeFactory in _testHostOrchestratorApplicationLifetimeFactories)
+        {
+            ITestHostOrchestratorApplicationLifetime service = testHostOrchestratorApplicationLifetimeFactory(serviceProvider);
+
+            // Check if we have already extensions of the same type with same id registered
+            if (lifetimes.Any(x => x.Uid == service.Uid))
+            {
+                ITestHostOrchestratorApplicationLifetime currentRegisteredExtension = lifetimes.Single(x => x.Uid == service.Uid);
```

Comment:
I am never a big fan of `Single` API because when it throws it's hard to debug, you don't know what was there.

Reply:
Same comment, it's a copy of https://github.com/microsoft/testfx/blob/b3b81f44e18c4f1c5acdbb3974791b3023e158d1/src/Platform/Microsoft.Testing.Platform/TestHost/TestHostManager.cs#L104-L109 😄